designs/validations.md

```diff
@@ -0,0 +1,288 @@
+Static data validations/defaults that only require one single API object **should** live in the `Validate` and `SetDefaults` api struct methods, thus in the `pkg/api/v1**` package. (*Note:* I'm more than open to relax this constraint if there is a good reason to, but given this is one of the only things that seem to resemble some kind consistency with regards to this whole topic, it seems worth to maintain it).
+
+Immutability validations have been treated differently (for some unknown reason) in the CLI than in webhooks. These checks **should** be implemented in the `api` package and **should not** be duplicated in two places. If the same checks are needed in both webhook and CLI, implement as standalone functions and reuse in both places. In the CLI, just read the object with a client and pass both old and new like the webhook does.
+
+The rest of validations and defaults (static data for a group of API structs and the context aware ones) **can** live anywhere, as long the dependency directions make sense, with a few constraints:
+* Context aware validation **cannot** live in `pkg/api/v1**`. This is to keep this package dependencies at a minimum.
+* **Avoid** packages like `validations`. Don't group validations for what they are, group them by what they validate (their context). For example, for docker validations, don't use `validations.ValidateDockerDesktopVersion`, think about `docker.ValidateDesktopVersion` in `pkg/docker`.
+* **Avoid** the `cmd` package. These tend to end up being untested and lack structure.
+
+Webhooks that only require static data validations/defaults **should** live in the `pkg/api/v1**` (where all present webhooks live today). If they have any other dependency the **cannot** live in the `api` package. (*Note*: we could possibly define a `webhooks` folder elsewhere if we think that could avoid problems in the future, leaving this up for discussion)
+
+### Scope (*what*)
+#### Defaults
+* Defaults should not **modify** data provided by the user, they can only **add** information. Any transformation needed should be moved to runtime operations and never get reflected back on the API objects.
+* They should have not secondary effects on anything else except the provided API object/s, including the internal state of the entity running the validation.
+* API objects should only be modified by the defaulting logic. This guarantees they are invariants during the CLI execution, allowing for workflow reentry/re-execution (think command fails and on the next execution we continue from where it stopped).
+
+#### Validations
+* They should have not secondary effects on **anything**, including the internal state of the entity running the validation.
+* API struct `Validate` methods should return a [`field.ErrorList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation/field#ErrorList). This facilitates using it from webhooks and ensures the API user facing error is as complete as possible.
+
+### Calling (*when*)
```

Comment:
Do we want to address the rare cases we have where some CLI flag parsing code wants needs to know what provider is in-use before deciding if a flag is required?

Reply:
Ok, if we can get an light consensus quickly on in, I'm down. here is my proposal (feel free to tweak wording):

> * In the special case where the code parsing/validating another type of CLI user input ( one that is not part of the EKS-A API, for example, command flags) requires information only available in the API (cluster config file):
>   1. Read the cluster config file
>   1. Parse the yaml content into a `cluster.Config`
>   1. Parse/validate the command flags
>   1. Construct a `cluster.Spec` with this `cluster.Config`
>   1. Continue with API defaults, validations and the rest of the command flow.
>   * Note how this requires complete decoupling between parsing the yaml content and the defaults and validations. Something that historically we haven't followed.

Reply:
It was just a thought I had. It can totally be addressed independently.